server/src/main/java/io/crate/execution/ddl/tables/TransportAddColumnAction.java

```diff
@@ -0,0 +1,370 @@
+
+        if (reference instanceof GeneratedReference genRef) {
+            // extract actual reference (geo or index) from the generated reference since we are using instance of below.
+            reference = genRef.reference();
+        }
+
+        String analyzer = null;
+        if (reference instanceof IndexReference indRef) {
+            analyzer = indRef.analyzer();
+        }
+
+        String geoTree = null;
+        GenericProperties geoProperties = null;
+        boolean geo = false;
+        if (reference instanceof GeoReference geoRef) {
+            geo = true;
+            geoTree = geoRef.geoTree();
+            Map<String, Object> geoMap = new HashMap<>();
+            if (geoRef.precision() != null) {
+                geoMap.put("precision", geoRef.precision());
+            }
+            if (geoRef.treeLevels() != null) {
+                geoMap.put("tree_levels", geoRef.treeLevels());
+            }
+            if (geoRef.distanceErrorPct() != null) {
+                geoMap.put("distance_error_pct", geoRef.distanceErrorPct());
+            }
+            if (geoMap.isEmpty() == false) {
+                geoProperties = new GenericProperties<>(geoMap);
+            }
+        }
+
+        var innerType = ArrayType.unnest(reference.valueType()); // In case of array reference, type options and type are specified based on the inner type.
+
+        addTypeOptions(columnProperties, innerType, geoProperties, geoTree, analyzer);
+        columnProperties.put("type", typeNameForESMapping(innerType, analyzer, reference.indexType() == IndexType.FULLTEXT));
+
+        columnProperties.put("position", reference.position());
+
+        if (reference.indexType() == IndexType.NONE) {
+            // we must use a boolean <p>false</p> and NO string "false", otherwise parser support for old indices will fail
+            columnProperties.put("index", false);
+        }
+
+        // We align with AnalyzedColumnDefinition.toMapping() but don't handle copy_to field since it's irrelevant for ADD COLUMN
+
+        if (reference.valueType().id() == ArrayType.ID) {
+            HashMap<String, Object> outerMapping = new HashMap<>();
+            outerMapping.put("type", "array");
+            if (ArrayType.unnest(reference.valueType()).id() == ObjectType.ID) {
+                objectMapping(columnProperties, reference, tree);
+            }
+            outerMapping.put("inner", columnProperties);
+            return outerMapping;
+        } else if (reference.valueType().id() == ObjectType.ID) {
+            objectMapping(columnProperties, reference, tree);
+        }
+
+        if (geo == false && reference.hasDocValues() == false) {
+            // We need to add this only if default value is true.
+            // Geo is always false but putting doc_values key to geo shape would cause MapperParsingException.
+            columnProperties.put(DOC_VALUES, "false");
+        }
```

Comment:
I feel like we need 3 states for `Reference.hasDocValues` - [_enabled_, _disabled_ and `not_supported`] to avoid such concrete checks  - if we add some other type without doc values (not likely but still) we would have to update this check as well,  `geo == false` looks a bit hacky. 

Before we used to compute mapping right in planner and semantics there was  "add only if different from default" We have there short circuit like `if (definition.storageProperties == null)  return` but when we stream a GeoRef we have to set `hasDocValues` and both true/false is not correct.

Reply:
If a type supports doc-values should probably become part of the `StorageSupport` property on the `DataType`. 
`Reference.hasDocValues` can then keep 2 states with the default already applied. That makes handling easier - otherwise we'd spread the logic to apply the default value everywhere.